### PR TITLE
[tests/core] Updated with BoringModel and added BoringDataModule

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of #4501
Updated `tests/core` with `BoringModel`. Added `BoringDataModule`.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/core/test_datamodules.py [hunk 1]
```diff
@@ -201,47 +199,54 @@ def test_train_loop_only(tmpdir):
...
+        max_epochs=1,
         weights_summary=None,
     )
 
     # fit model
     result = trainer.fit(model, dm)
-    assert result == 1
-    assert trainer.logger_connector.callback_metrics['loss'] < 0.6
+    assert result
+    # assert trainer.callback_metrics['train_loss'] < 0.6
```
Comment (rohitgr7) [inline tests/core/test_datamodules.py:231]: what should be done with such assertions when using `BoringModel`? Since the dataset used with `BoringModel` is completely random.

Comment (awaelchli) [reply tests/core/test_datamodules.py:231]: maybe we need a BoringMNISTModel for that 🤣

File: tests/base/boring_model.py [hunk 2]
```diff
@@ -122,10 +125,48 @@ def configure_optimizers(self):
...
+        if stage == "fit" or stage is None:
+            self.random_train = Subset(self.random_full, indices=range(64))
+            self.random_val = Subset(self.random_full, indices=range(64, 128))
+            self.dims = self.random_train[0].shape
+
+        if stage == "test" or stage is None:
+            self.random_test = Subset(self.random_full, indices=range(128, 192))
+            self.dims = getattr(self, "dims", self.random_test[0].shape)
+
+        self.non_picklable = lambda x: x ** 2
```
Comment (awaelchli) [inline tests/base/boring_model.py:157]: I think the tests that require `non_picklable` should subclass and override setup(), just as we do for BoringModel

File: tests/base/boring_model.py [hunk 3]
```diff
@@ -122,10 +125,48 @@ def configure_optimizers(self):
...
+        return DataLoader(self.random_val)
+
+    def test_dataloader(self):
+        return DataLoader(self.random_test)
+
+    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        checkpoint[self.__class__.__name__] = self.__class__.__name__
+
+    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+        self.checkpoint_state = checkpoint.get(self.__class__.__name__)
```
Comment (awaelchli) [inline tests/base/boring_model.py:172]: where is it used?

Comment (rohitgr7) [reply tests/base/boring_model.py]: updated.

File: tests/base/boring_model.py (hunk 2 above)
Comment (rohitgr7) [reply tests/base/boring_model.py]: updated.

**Review by Borda — APPROVED** [review]

File: tests/core/test_datamodules.py (hunk 1 above)
Comment (Borda) [reply tests/core/test_datamodules.py:232]: we shall replace result validation with one end-to-end test..

**Review by tchaton — APPROVED** [review]

File: tests/core/test_datamodules.py [hunk 4]
```diff
@@ -201,47 +200,61 @@ def test_train_loop_only(tmpdir):
...
```
Comment (justusschock) [inline tests/core/test_datamodules.py:232]:
```suggestion
    # TODO: add end-to-end test for: assert trainer.callback_metrics['train_loss'] < 0.6
```

File: tests/core/test_datamodules.py [hunk 5]
```diff
@@ -290,20 +301,22 @@ def test_full_loop(tmpdir):
 
     # test
     result = trainer.test(datamodule=dm)
-    result = result[0]
-    assert result['test_acc'] > 0.8
+    # result = result[0]
+    # assert result['test_acc'] > 0.8
```
Comment (justusschock) [inline tests/core/test_datamodules.py:305]:
```suggestion
    # TODO: add end-to-end test for: assert result[0]['test_acc'] > 0.8
```

File: tests/core/test_datamodules.py [hunk 6]
```diff
@@ -319,43 +332,41 @@ def test_trainer_attached_to_dm(tmpdir):
...
     result = trainer.fit(model, dm)
-    assert result == 1
+    assert result
 
     # test
     result = trainer.test(datamodule=dm)
-    result = result[0]
-    assert result['test_acc'] > 0.8
+    # result = result[0]
+    # assert result['test_acc'] > 0.8
```
Comment (justusschock) [inline tests/core/test_datamodules.py:357]:
```suggestion
    # TODO: Add end-to-end test for: assert result[0]['test_acc'] > 0.8
```

File: tests/core/test_datamodules.py [hunk 7]
```diff
@@ -364,18 +375,17 @@ def test_full_loop_dp(tmpdir):
...
```
Comment (justusschock) [inline tests/core/test_datamodules.py:383]:
```suggestion
    # TODO: Add end-to-end test for: assert result[0]['test_acc'] > 0.8
```

File: tests/core/test_datamodules.py [hunk 8]
```diff
@@ -201,47 +200,61 @@ def test_train_loop_only(tmpdir):
...
+        max_epochs=1,
         weights_summary=None,
     )
 
     # fit model
     result = trainer.fit(model, dm)
-    assert result == 1
-    assert trainer.logger_connector.callback_metrics['loss'] < 0.6
+    assert result
+    # assert trainer.callback_metrics['loss'] < 0.6
```
Comment (justusschock) [inline tests/core/test_datamodules.py:210]:
```suggestion
    # TODO: Add end-to-end test for:  assert trainer.callback_metrics['loss'] < 0.6
```

**Review by justusschock — APPROVED** [review]

**Comment by rohitgr7** [comment]
@williamFalcon mind review?

**Review by williamFalcon — APPROVED** [review]
